### PR TITLE
ES7 `async`/`await` implementation with concise syntax, plus tests

### PR DESCRIPTION
**Pull request by gkovacs**
Based on https://github.com/gkz/LiveScript/pull/836 which unfortunately wasn't merged because it lacked tests and the original author (@summivox) asked for it to be taken over, please see there for background and discussion.

I have added tests in this pull request, I have been using this in using this in production for 3 months without issue so I am fairly certain it is correct code-wise but let me know if there are any additional tests that should be added.

It appears the documentation for this project is maintained in a separate branch (gh-pages, I believe) so I will make a pull request to that branch adding documentation for this feature once this pull request is merged.

### CODE REVIEW

**Comment by gkovacs** [comment]
Note the reason the build says it is failing on nodejs 4 and 6 is because those versions don't support the async/await feature, it is only supported in nodejs 7 and 8. Should I edit travis.yml accordingly (remove nodejs 4 and 6 from testing, add nodejs 8)?

File: test/async.ls [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+x = ->>
+  2 + await do ->>
+    6 + await do ->>
+      5 + await do ->>
+        3
+
+y <- x!then
+
+eq y, 16
```
Comment (rhendric) [inline test/async.ls]: Please add one test each for:
* `async function name(arglist)` syntax
* `!->>` (hushed async functions should still return Promises, but they shouldn't be implicitly fulfilled with the value of the body)
* `-->>`
* `~>>`
* a `compile-throws` test for `->>*`

No need to do all the combinations (`!~~>>`), just a quick smoke test of each is fine.

File: src/lexer.ls [hunk 2]
```diff
@@ -240,6 +240,9 @@ exports <<<
                 or last.1 is 'import' and 'All'
                     last.1 += that
                     return 3
+                if last.0 is 'yield' and last.1 is 'await'
+                    last.1 += 'all'
+                    return 3
```
[the remark this answers is not in the document]
Comment (gkovacs) [reply src/lexer.ls]: Not needed, removed it

File: src/ast.ls [hunk 3]
```diff
@@ -1183,21 +1186,21 @@ class exports.Yield extends Node
 
     children: <[ it ]>
 
-    show: -> if @op is 'yieldfrom' then 'from' else ''
+    show: -> switch @op
+    | 'yield' => ''
+    | 'yieldfrom' => 'from'
+    | 'await' => 'await'
+    | _ => ''
```
[the remark this answers is not in the document]
Comment (gkovacs) [reply src/ast.ls]: Done

File: test/async.ls [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+ok ->> 1
+
+# named async function
+ok <| :fn ->> 2
+
+# async function definition
+x = ->>
+  3
+
+y <- x!then
```
[the remark this answers is not in the document]
Comment (gkovacs) [reply test/async.ls]: Done

File: test/async.ls (hunk 1 above)
Comment (gkovacs) [reply test/async.ls]: Done. The ->>* results in a parse error which I didn't want to over-fit the test to (also I couldn't find a straightforward way to detect ->>* without complicating the grammar) so I added compile-parse-error  method that looks for parse errors in scripts/test.

**Comment by gkovacs** [comment]
@rhendric OK removed the dead code and added those additional tests, should be ready for another round of reviews

**Comment by rhendric** [comment]
I manually merged to master, cleaning up the git history and making a few tweaks mostly to the testing. Thank you again!

**Comment by ymeine** [comment]
This is a great feature, finally! :)

I was starting to change my code, and there is still one thing I am missing though: async backcalls. For now it's trying to parse it as function composition. This is mainly to simulate top-level await (in a more controlled way), but it could also be useful when using any function receiving an async callback.

Is there any plan to implement the async backcall syntax?

**Comment by rhendric** [comment]
@ymeine, I believe that the original thinking around leaving them out was that backcalls are the poor man's version of await, and mixing both syntax features would be unnecessarily confusing and shouldn't be encouraged. I don't think they'd be hard to add, but I'd like to see more of an argument for why they're needed (instead of promisifying any callback-based functions and only using await instead of backcalls).

**Comment by rhendric** [comment]
If what you really want is top-level await, I think I'd rather try to give you top-level await. LiveScript already compiles everything inside an IIFE by default; I can look into how what it would take to make that IIFE async if there are any awaits outside of any functions. (It would be a compile error to try to use top-level await with the `--bare` compilation option.) How does that sound?

**Comment by ymeine** [comment]
I think that's a good trade-off, it would cover my main use case, while still being a non-intrusive feature (applied only if there is a top-level await) and taking advantage of an already present feature (the wrapping IIFE). Moreover it would go beyond and solve my issue which was the auto-invocation of the backcall.
To me that sounds great!

**Comment by bartosz-m** [comment]
@ymeine You could try two packages that I've coded today:
[transform-top-level-await](https://www.npmjs.com/package/livescript-transform-top-level-await)  it allows to write
``` livescript
foo = Promise.resolve \foo
bar = Promise.resolve \bar 
foo-bar = ->> "#{await foo} #{await bar}" 
x = await foo-bar!
```
add  [transform-implicit-async](https://www.npmjs.com/package/livescript-transform-implicit-async) and you can omit `->>` and just write `->`
``` livescript
foo = Promise.resolve \foo
bar = Promise.resolve \bar 
foo-bar = -> "#{await foo} #{await bar}" 
x = await foo-bar!
```

**Pull request closed without merging**